### ISSUE TITLE
Intermittent error when pulling from the repository.

### ISSUE THREAD

**Issue by rfeng2023** [rfeng2023/mmcloud#120]
We encountered the following error during a `git pull` operation:  
```
fatal: Failed to checksum '.git/objects/pack/tmp_pack_TEVMwv': No such file or directory
fatal: fetch-pack: invalid index-pack output
```

This issue does not occur consistently but happens fairly often, especially when there are significant changes in the upstream repository that could lead to conflicts.

**Comment by Ashley-Tung** [comment]
With the new setup, can we close this? We can open it back up if this appears again. Thank you!

**Comment by gaow** [comment]
This is still an issue. 

I wonder if we can try this on the EBS (or whatever name it is) system where we install software, to see if it helps? If so, we should update our script to mount automatically a folder just for github repo clone purpose.

**Comment by Ashley-Tung** [comment]
After some digging around in documentation and feedback, it turns out git does not work well with S3 for the issue of Eventual Consistency:
- S3 is an object store with eventual consistency, meaning that changes (like new files created in .git/objects) might not immediately be visible, causing errors when Git attempts to verify or use them. This can lead to issues when `git pulling`, when versions are inconsistent between what is seen locally and what is actually on the S3. You can think of it as S3 does not immediately update its contents, there is some small time difference (big enough to affect git)
- This means for files that rely on versioning and time integrity, s3fs does not guarantee it. Just fine for just storing data and package but bad for git
- There can also be cases where the same user could have different Vms open and testing the repo on two different machines, which can also lead to versioning issues

So what can we do? Since all the data is on S3, we would still want some kind of cloud storage for the repo so that we can access the data. We would need a file syste that does guarantee data integrity and consistency for concurrent writing, which EFS does guarantee. JuiceFS + S3 also guarantees this and we shouldn't have to worry about slowness since we would be using this for a git repo storage instead of packages.

For EFS, I am thinking in the interactive setup, under each user's directory, we could have a subdir that would hold all the repos. It would be symlinked to `/home/ubuntu` and users can work on their repos there

I am happy to discuss this in further detail as well, @gaow . I think it would also be a good idea to loop in @danielnachun

**Comment by danielnachun** [comment]
We could standardize how we pull repos using `ghq`: https://github.com/x-motemen/ghq.  I exclusively clone `git` repos with that tool, which stores all the repos in the folder `${HOME}/ghq` with the following structure:
```
ghq
  github.com
    user1
      repo1
      repo2
    user2
      repo1
  gitlab.com
    user1
      repo1
```
Repos are cloned with `ghq get URL` so for this repo it would be `ghq get https://github.com/rfeng2023/mmcloud`.

@Ashley-Tung if we standardized on this setup, we'd just need to create symlink for `/home/ubuntu/ghq` to EFS just as we've done for `pixi`, which has worked very well for us.  We'd just need to document for people that they need to use that command instead of the usual `git clone` when cloning repositories.

**Comment by Ashley-Tung** [comment]
I am good with this solution if @gaow is as well! I will document this in the wiki once we've implemented and tested this as well

**Comment by danielnachun** [comment]
Two details I forgot to mention above:

The package name on conda-forge is `go-ghq`, so it would be installed with `pixi global install go-ghq`.

There is an API interface for GitHub that can be used to directly download a git repo at a specific commit without needing to use `git` at all.  To download the a tarball of the code for this repo at the latest commit, you can `wget` or `curl` with this URL https://github.com/rfeng2023/mmcloud/archive/8c664701a074ebfd514980618920f2fbe9d80af2.tar.gz.

If you use this approach, the downloaded archive is *not* a git repository but simply a copy of the files in the state they were in that commit, so you can commit changes, push, pull etc.  This is obviously not of much use for interactive development, but might be use for batch jobs where the code is not expected to change.  You can also use a tag for this of course, which is better for reproducibility if it makes sense to do so.

**Comment by Ashley-Tung** [comment]
Do you think `go-ghq` should be a shared package? If so, I can get a PR ready to implement this

**Comment by danielnachun** [comment]
No, let's add it to the setup script in Gao's repo.  It's a packages users need to be able to update on their own.

**Comment by danielnachun** [comment]
Existing setups will just need to run `pixi global install go-ghq` to add it.

**Comment by Ashley-Tung** [comment]
I see, let me repeat to make sure I understood correctly:
- We'll put `pixi global install go-ghq` in the `pixi-mamba` script for new users to install but for existing setups, we can let them know (an update the wiki) to run the command themselves
- On the `host_init` interactive setup, we can create a GIT subdir (name pending) under their user directory. This will be symlinked to `/home/ubuntu` so they will access it normally on the interactive setup
- We would need to let them know to use `ghq get` to clone (would also be in the wiki) and then they can use git normally after that?

**Comment by Ashley-Tung** [comment]
https://github.com/rfeng2023/mmcloud/pull/133

**Comment by Ashley-Tung** [comment]
Issue #120 should be good to go now. We can leave it open in case we hit the same error

Wiki update: https://github.com/gaow/lab-wiki/pull/56
